### PR TITLE
 feat(pm_list): add required fields for sofort

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Resolves [#3704](https://github.com/juspay/hyperswitch-cloud/issues/3704)
### Test Cases
1. Create a sofort payment with mollie with empty payment method object. The payment should succeed. 

```
{
  "amount": 1000,
  "currency": "EUR",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "amount_to_capture": 1000,
  "customer_id": "StripeCustomer",
  "email": "abcdef123@gmail.com",
  "name": "John Doe",
  "phone": "999999999",
  "phone_country_code": "+65",
  "description": "Its my first payment request",
  "authentication_type": "three_ds",
  "return_url": "https://google.com",
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "payment_method": "bank_redirect",
  "payment_method_type": "sofort",
  "payment_method_data": {
    "bank_redirect": {
      "sofort": {
      }
    }
  }
}
```

Must be tested for each of this connectors

1. Adyen
```
None
```

2. ACI
```
Required fields: 
     payment_method_data.bank_redirect.sofort.country
```
3. GlobalPayments
```
Required fields: 
    billing.address.country
```
4. Mollie
```
None
```
5. Nuvie
```
Required fields: 
     email
     billing.address.first_name
     billing.address.last_name
     billing.address.country
```
6. Paypal 
```
Required fields: 
     payment_method_data.bank_redirect.sofort.billing_details.billing_name
     payment_method_data.bank_redirect.sofort.country
```
7. Shift4
```
None
```
8. Stripe  
```
Required fields for mandates
     payment_method_data.bank_redirect.sofort.billing_details.billing_name
     payment_method_data.bank_redirect.sofort.billing_details.email
```
9. Trustpay
```
Required fields: 
     billing.address.first_name
     billing.address.line1
     billing.address.city
     billing.address.zip
     billing.address.country
```
10. Nexinets
```
None
```

-> Create a payment with confirm `false`

-> list payment methods, with client secret and publishable key

```
curl --location 'http://localhost:8080/account/payment_methods?client_secret={{client_secret}}' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key:{{api_key}}' \
--data ''
```
 Required fields must be populated, according to the connector

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/configs/defaults.rs [hunk 1]
```diff
@@ -4460,14 +4460,324 @@ impl Default for super::settings::RequiredFields {
...
+                                                }
+                                            ),
+                                            (
+                                                "billing.address.country".to_string(),
+                                                RequiredFieldInfo {
+                                                    required_field: "billing.address.country".to_string(),
+                                                    display_name: "country".to_string(),
+                                                    field_type: enums::FieldType::UserAddressCountry{
+                                                        options: vec![
+                                                            "AT".to_string(),
```
Comment (prasunna09) [inline crates/router/src/configs/defaults.rs:4589]: this is the supported countries for Sofort pm via nuvei, - https://nuvei.com/apm/sofort/
is there any reason why these countries are not added? 
Please check for other connectors docs and update the same

**Review by prasunna09 — CHANGES_REQUESTED** [review]

File: crates/router/src/configs/defaults.rs (hunk 1 above)
Comment (AkshayaFoiger) [reply crates/router/src/configs/defaults.rs]: These countries must be added, missed it by mistake. Will update it

File: config/config.example.toml [hunk 2]
```diff
@@ -380,7 +380,8 @@ pay_later.klarna = { connector_list = "adyen" }     # Mandate supported payment
 bank_debit.ach = { connector_list = "gocardless" }  # Mandate supported payment method type and connector for bank_debit
 bank_debit.becs = { connector_list = "gocardless" } # Mandate supported payment method type and connector for bank_debit
 bank_debit.sepa = { connector_list = "gocardless" } # Mandate supported payment method type and connector for bank_debit
-bank_redirect.ideal = {connector_list = "stripe,adyen"} # Mandate supported payment method type and connector for bank_redirect
+bank_redirect.ideal = {connector_list = "stripe,adyen,globalpay"} # Mandate supported payment method type and connector for bank_redirect
```
Comment (prasunna09) [inline config/config.example.toml:383]: why are ideal changes in sofort pr?

File: crates/router/src/connector/adyen/transformers.rs [hunk 3]
```diff
@@ -2751,10 +2751,7 @@ fn get_redirect_extra_details(
                     country,
                     preferred_language,
                     ..
-                } => Ok((
-                    Some(preferred_language.to_string()),
-                    Some(country.to_owned()),
-                )),
+                } => Ok((preferred_language.clone(), country.clone())),
```
Comment (prasunna09) [inline crates/router/src/connector/adyen/transformers.rs:2754]: is this not mandatory now?? earlier it was a mandatory field ryt

File: crates/router/src/connector/paypal/transformers.rs [hunk 4]
```diff
@@ -360,8 +360,15 @@ fn get_payment_source(
             preferred_language: _,
             billing_details,
         } => Ok(PaymentSourceItem::Sofort(RedirectRequest {
-            name: billing_details.get_billing_name()?,
-            country_code: *country,
+            name: billing_details
+                .clone()
+                .ok_or(errors::ConnectorError::MissingRequiredField {
+                    field_name: "sofort.billing_details.name",
```
Comment (prasunna09) [inline crates/router/src/connector/paypal/transformers.rs:366]:
```suggestion
                .ok_or(errors::ConnectorError::MissingRequiredField {
                    field_name: "sofort.billing_details",
```

File: crates/router/src/connector/stripe/transformers.rs [hunk 5]
```diff
@@ -290,9 +290,9 @@ pub struct StripeSofort {
     #[serde(rename = "payment_method_data[type]")]
     pub payment_method_data_type: StripePaymentMethodType,
     #[serde(rename = "payment_method_options[sofort][preferred_language]")]
-    preferred_language: String,
+    preferred_language: Option<String>,
     #[serde(rename = "payment_method_data[sofort][country]")]
-    country: api_enums::CountryAlpha2,
+    country: Option<api_enums::CountryAlpha2>,
```
Comment (prasunna09) [inline crates/router/src/connector/stripe/transformers.rs:295]: is this not mandatory??

File: crates/router/src/connector/stripe/transformers.rs [hunk 6]
```diff
@@ -1115,36 +1115,10 @@ impl TryFrom<(&payments::BankRedirectData, Option<bool>)> for StripeBillingAddre
...
-                        name: billing_name,
-                        email: billing_email,
-                        ..Self::default()
-                    }),
-                }
-            }
+            } => Ok(get_stripe_sepa_dd_mandate_billing_details(
+                billing_details,
+                is_customer_initiated_mandate_payment,
+            )?),
```
Comment (prasunna09) [inline crates/router/src/connector/stripe/transformers.rs:1121]: why are ideal changes done in sofort pr?

**Review by prasunna09 — CHANGES_REQUESTED** [review]

File: config/config.example.toml (hunk 2 above)
Comment (AkshayaFoiger) [reply config/config.example.toml:383]: Globalpay support mandates for ideal. But it was missed in the ideal PR. Hence, this PR also contains minor config changes of `ideal`

File: crates/router/src/connector/adyen/transformers.rs (hunk 3 above)
Comment (AkshayaFoiger) [reply crates/router/src/connector/adyen/transformers.rs:2754]: No country and preferred_language is not a [required field](https://docs.adyen.com/payment-methods/sofort/api-only/#make-payment)

File: crates/router/src/connector/stripe/transformers.rs (hunk 5 above)
Comment (AkshayaFoiger) [reply crates/router/src/connector/stripe/transformers.rs:295]: No country is not mandatory for [stripe](https://stripe.com/docs/payments/sofort/accept-a-payment?platform=web&ui=API#create-payment-intent)

File: crates/router/src/connector/stripe/transformers.rs (hunk 6 above)
Comment (AkshayaFoiger) [reply crates/router/src/connector/stripe/transformers.rs:1121]: Recurring pay for Ideal and Sofort in stripe happens through SEPA Direct Debit payments. Therefore both requires same request body resulting in repeated code, hence I removed the redundancy by moving it into a function get_stripe_sepa_dd_mandate_billing_details

File: crates/router/src/connector/stripe/transformers.rs (hunk 5 above)
Comment (prasunna09) [reply crates/router/src/connector/stripe/transformers.rs:295]: as we checked , it is mandatory. make the necessary changes

File: crates/router/src/connector/stripe/transformers.rs (hunk 6 above)
Comment (prasunna09) [reply crates/router/src/connector/stripe/transformers.rs:1121]: can u please add the same comment in code to avoid confusion in future

**Review by prasunna09 — CHANGES_REQUESTED** [review]

File: crates/router/src/configs/defaults.rs [hunk 7]
```diff
@@ -4748,14 +4748,285 @@ impl Default for super::settings::RequiredFields {
...
+                                                },
+                                                value: None,
+                                            }
+                                        ),
+                                        (
+                                            "payment_method_data.bank_redirect.sofort.billing_details.billing_name".to_string(),
+                                            RequiredFieldInfo {
+                                                required_field: "payment_method_data.bank_redirect.sofort.billing_details.billing_name".to_string(),
+                                                display_name: "billing_name".to_string(),
+                                                field_type: enums::FieldType::UserFullName,
```
Comment (prasunna09) [inline crates/router/src/configs/defaults.rs:4917]:
```suggestion
                                                field_type: enums::FieldType::UserBillingName,
```

**Review by prasunna09 — CHANGES_REQUESTED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Pull request merged**